libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp

```diff
@@ -114,9 +115,44 @@ void DagBlockPacketHandler::sendBlock(dev::p2p::NodeID const &peer_id, taraxa::D
   }
 }
 
-void DagBlockPacketHandler::onNewBlockReceived(DagBlock &&block) {
+void DagBlockPacketHandler::onNewBlockReceived(DagBlock &&block, const std::shared_ptr<TaraxaPeer> &peer) {
   if (dag_blk_mgr_) {
-    dag_blk_mgr_->insertAndVerifyBlock(std::move(block));
+    const auto verified = dag_blk_mgr_->insertAndVerifyBlock(std::move(block));
+    switch (verified) {
+      case DagBlockManager::InvalidBlock:
+      case DagBlockManager::FailedVdfVerification:
+      case DagBlockManager::NotEligible:
+        LOG(log_er_) << "DagBlock" << block.getHash() << " failed verification with error code " << verified
+                     << " . Peer " << peer->getId() << " will be disconnected and marked as malicious.";
+        disconnect(peer->getId(), dev::p2p::UserReason);
+        peers_state_->set_peer_malicious(peer->getId());
+        break;
```

Comment:
every code block that does disconnect + setting per malicious should be replaced by:
`throw MaliciousPeerException(err_msg);`

Reply:
Fixed